creator-node/src/services/stateMachineManager/clockUtils.js

```diff
@@ -0,0 +1,130 @@
+   * Also returns a set of any nodes that were unhealthy when queried for clock values.
+   * @param {Object} replicaSetNodesToUserWalletsMap map of <replica set node : wallets>
+   *
+   * @returns {Object} { replicasToUserClockStatusMap: map(replica node => map(wallet => clockValue)), unhealthyPeers: Set<string> }
+   */
+  static async retrieveClockStatusesForUsersAcrossReplicaSet(
+    replicasToWalletsMap
+  ) {
+    const replicasToUserClockStatusMap = {}
+    const unhealthyPeers = new Set()
+
+    /** In parallel for every replica, fetch clock status for all users on that replica */
+    const replicas = Object.keys(replicasToWalletsMap)
+    await Promise.all(
+      replicas.map(async (replica) => {
+        replicasToUserClockStatusMap[replica] = {}
+
+        const walletsOnReplica = replicasToWalletsMap[replica]
+
+        // Make requests in batches, sequentially, to ensure POST request body does not exceed max size
+        for (
+          let i = 0;
+          i < walletsOnReplica.length;
+          i += MAX_BATCH_CLOCK_STATUS_BATCH_SIZE
+        ) {
+          const walletsOnReplicaSlice = walletsOnReplica.slice(
+            i,
+            i + MAX_BATCH_CLOCK_STATUS_BATCH_SIZE
+          )
+
+          const axiosReqParams = {
+            baseURL: replica,
+            url: '/users/batch_clock_status',
+            method: 'post',
+            data: { walletPublicKeys: walletsOnReplicaSlice },
+            timeout: BATCH_CLOCK_STATUS_REQUEST_TIMEOUT
+          }
+
+          // Sign request to other CN to bypass rate limiting
+          const { timestamp, signature } = generateTimestampAndSignature(
+            { spID: SP_ID },
+            DELEGATE_PRIVATE_KEY
+          )
+          axiosReqParams.params = { spID: SP_ID, timestamp, signature }
+
+          let batchClockStatusResp = []
+          let errorMsg
+          try {
+            batchClockStatusResp = (
+              await retry(async () => axios(axiosReqParams), {
+                retries: MAX_USER_BATCH_CLOCK_FETCH_RETRIES
+              })
+            ).data.data.users
+          } catch (e) {
+            errorMsg = e
+          }
+
+          // If failed to get response after all attempts, add replica to `unhealthyPeers` set for reconfig
+          if (errorMsg) {
+            logger.error(
+              `retrieveClockStatusesForUsersAcrossReplicaSet Could not fetch clock values for wallets=${walletsOnReplica} on replica=${replica} ${errorMsg.toString()}`
+            )
+            unhealthyPeers.add(replica)
+          }
+
+          // Add batch response data to aggregate output map
+          batchClockStatusResp.forEach((userClockValueResp) => {
+            const { walletPublicKey, clock } = userClockValueResp
+            replicasToUserClockStatusMap[replica][walletPublicKey] = clock
+          })
+        }
+      })
+    )
+
+    return {
+      replicasToUserClockStatusMap,
+      unhealthyPeers
+    }
+  }
```

Comment:
Copy/pasted from [here](https://github.com/AudiusProject/audius-protocol/blob/4c559fd7b2a7e3aa5e4772c5b153260841fe9072/creator-node/src/snapbackSM/snapbackSM.js#L814-L886). Changes:
- logger (use logger instead of this.log)
- import location of constants
- spId fetched directly from config instead of class variable

Reply:
don't love each submodule creating its own logger bc we don't get any consistency
i'm not 100% on the best pattern but somehow would be nice to have each submodule create a child logger from the parent and progressively add prefixes, that way all logs inside stateMachineManager have `stateMachineManager:<submodule>:<submodule>...` prefixes

would love input here tho

Reply:
synced about how this should be json property rather than in the string. made Linear ticket to track this: https://linear.app/audius/issue/CON-167/make-state-machine-use-consistent-logging-pattern